File: receiver.py
import network
import time
import json
import socket
import ssl
import os
from machine import Pin

# ============ CONFIGURATION ============
WIFI_SSID     = "Hotspot"
WIFI_PASSWORD = "12345678"
WS_HOST       = "holy-backend-64n1.onrender.com"
WS_PORT       = 443
DEVICE_ID     = "pcb1"

# ============ DIRECTION CODES ============
D_BASE_LEFT  = 0
D_BASE_RIGHT = 1
D_LP_UP      = 2
D_LP_DOWN    = 3
D_RP_UP      = 4
D_RP_DOWN    = 5

DIR_NAMES = {
    0: "BASE_L",
    1: "BASE_R",
    2: "LP_UP",
    3: "LP_DN",
    4: "RP_UP",
    5: "RP_DN",
}

# ============ RELAY PINS ============
# One GPIO per direction → connects to relay module input
# GP2=Base Left, GP3=Base Right, GP4=LP Up, GP5=LP Down, GP6=RP Up, GP7=RP Down
relay_pins = {
    D_BASE_LEFT:  Pin(2, Pin.OUT),
    D_BASE_RIGHT: Pin(3, Pin.OUT),
    D_LP_UP:      Pin(4, Pin.OUT),
    D_LP_DOWN:    Pin(5, Pin.OUT),
    D_RP_UP:      Pin(6, Pin.OUT),
    D_RP_DOWN:    Pin(7, Pin.OUT),
}

# Conflict pairs — only one per pair can be HIGH at a time (last wins)
CONFLICT_PAIRS = [
    (D_BASE_LEFT,  D_BASE_RIGHT),
    (D_LP_UP,      D_LP_DOWN),
    (D_RP_UP,      D_RP_DOWN),
]

# All relays OFF at boot
for p in relay_pins.values():
    p.value(0)

# ============ COMMAND HANDLER ============
def handle(msg):
    d = msg.get('d')
    s = msg.get('s')

    if d is None or s is None:
        return
    if d not in relay_pins:
        print(f"  ? Unknown d: {d}")
        return

    name = DIR_NAMES.get(d, str(d))

    if s == 1:
        # Turn off conflicting pin first
        for pair in CONFLICT_PAIRS:
            if d in pair:
                other = pair[1] if pair[0] == d else pair[0]
                if relay_pins[other].value() == 1:
                    relay_pins[other].value(0)
                    print(f"  ✗ Cleared conflict: {DIR_NAMES[other]}")
                break
        relay_pins[d].value(1)
        print(f"  ▶ {name} ON")

    else:
        relay_pins[d].value(0)
        print(f"  ■ {name} OFF")

def stop_all():
    for p in relay_pins.values():
        p.value(0)
    print("  All relays OFF")

# ============ WIFI ============
def connect_wifi():
    wlan = network.WLAN(network.STA_IF)
    wlan.active(True)
    wlan.connect(WIFI_SSID, WIFI_PASSWORD)
    print("Connecting to WiFi...")
    timeout = 10
    while not wlan.isconnected() and timeout > 0:
        time.sleep(1)
        timeout -= 1
    if wlan.isconnected():
        ip, subnet, gateway, _ = wlan.ifconfig()
        wlan.ifconfig((ip, subnet, gateway, '8.8.8.8'))
        print(f"✓ WiFi connected! IP: {ip}\n")
        return True
    print("✗ WiFi failed")
    return False

# ============ WEBSOCKET ============
def _ws_send(sock, text):
    msg_bytes = text.encode('utf-8')
    mask = os.urandom(4)
    frame = bytearray([0x81, 0x80 | len(msg_bytes)])
    frame.extend(mask)
    for i, byte in enumerate(msg_bytes):
        frame.append(byte ^ mask[i % 4])
    sock.send(bytes(frame))

def connect_websocket():
    import gc
    gc.collect()
    try:
        print("Connecting to WebSocket...")
        addr = socket.getaddrinfo(WS_HOST, WS_PORT)[0][-1]
        print(f"Resolved: {addr}")

        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.connect(addr)
        sock = ssl.wrap_socket(sock, server_hostname=WS_HOST)
        print("✓ SSL connected")

        # WebSocket handshake
        sock.send((
            "GET / HTTP/1.1\r\n"
            f"Host: {WS_HOST}\r\n"
            "Upgrade: websocket\r\n"
            "Connection: Upgrade\r\n"
            "Sec-WebSocket-Key: dGhlIHNhbXBsZSBub25jZQ==\r\n"
            "Sec-WebSocket-Version: 13\r\n\r\n"
        ).encode())

        response = sock.recv(1024).decode()
        if "101" not in response:
            print(f"✗ Handshake failed: {response[:80]}")
            sock.close()
            return None
        print("✓ WebSocket handshake OK")

        # Register
        _ws_send(sock, json.dumps({
            "type":       "register",
            "clientType": "pcb",
            "deviceId":   DEVICE_ID
        }))
        print(f"✓ Registered as [{DEVICE_ID}]\n")
        return sock

    except Exception as e:
        print(f"✗ WS error: {e}")
        import sys
        sys.print_exception(e)
        return None

def recv_msg(sock):
    try:
        h = sock.recv(2)
        if len(h) < 2:
            return None
        l = h[1] & 0x7F
        if l == 126: l = int.from_bytes(sock.recv(2), 'big')
        elif l == 127: l = int.from_bytes(sock.recv(8), 'big')
        p = sock.recv(l)
        return json.loads(p.decode()) if p else None
    except:
        return None

# ============ MAIN ============
print("=" * 40)
print("  Spray Bot - Receiver (Relay Mode)")
print("=" * 40 + "\n")

if not connect_wifi():
    print("Exiting.")
else:
    while True:
        sock = connect_websocket()
        if not sock:
            print("Retry in 5s...")
            time.sleep(5)
            continue

        print("Listening for commands...\n")

        try:
            while True:
                msg = recv_msg(sock)
                if msg is None:
                    print("Connection lost.")
                    break
                if msg.get('type') == 'command' or 'd' in msg:
                    t = time.localtime()
                    print(f"[{t[3]:02d}:{t[4]:02d}:{t[5]:02d}] d={msg.get('d')} s={msg.get('s')}")
                    handle(msg)

        except Exception as e:
            print(f"Error: {e}")
        finally:
            stop_all()
            try: sock.close()
            except: pass
            time.sleep(5)

File: receiver.cpp
/*
  Spray Bot - Receiver 1 (pcb1)
  Change DEVICE_ID only for pcb2 and pcb3 — everything else identical
*/

#include <Arduino.h>
#include <WiFi.h>
#include <WebSockets.h>
#include <WebSocketsClient.h>
#include <ArduinoJson.h>

// ============ CONFIGURATION ============
const char* WIFI_SSID     = "test";
const char* WIFI_PASSWORD = "12345678";
const char* WS_HOST       = "holy-backend-64n1.onrender.com";
const int   WS_PORT       = 443;
const char* WS_PATH       = "/";
const char* DEVICE_ID     = "pcb1";   // ← Change to "pcb2" or "pcb3" for other receivers

// ============ DIRECTION CODES ============
#define D_BASE_LEFT  0
#define D_BASE_RIGHT 1
#define D_LP_UP      2
#define D_LP_DOWN    3
#define D_RP_UP      4
#define D_RP_DOWN    5
#define DIR_COUNT    6

const char* DIR_NAMES[DIR_COUNT] = {
  "BASE_L", "BASE_R", "LP_UP", "LP_DN", "RP_UP", "RP_DN"
};

// ============ RELAY PINS ============
const int RELAY_PINS[DIR_COUNT] = { 2, 3, 4, 5, 6, 7 };
// NOTE: Active-low relay module — HIGH = OFF, LOW = ON

// ============ CONFLICT PAIRS ============
const int CONFLICT_PAIRS[3][2] = {
  { D_BASE_LEFT,  D_BASE_RIGHT },
  { D_LP_UP,      D_LP_DOWN    },
  { D_RP_UP,      D_RP_DOWN    }
};

WebSocketsClient wsClient;
bool wsConnected = false;

unsigned long lastWifiCheck = 0;
#define WIFI_CHECK_INTERVAL 10000

// ============ RELAY CONTROL ============
void stopAll() {
  for (int i = 0; i < DIR_COUNT; i++)
    digitalWrite(RELAY_PINS[i], HIGH);
  Serial.println("  All relays OFF");
}

void handleCommand(int d, int s) {
  if (d < 0 || d >= DIR_COUNT) return;

  if (s == 1) {
    for (int i = 0; i < 3; i++) {
      if (CONFLICT_PAIRS[i][0] == d || CONFLICT_PAIRS[i][1] == d) {
        int other = (CONFLICT_PAIRS[i][0] == d) ? CONFLICT_PAIRS[i][1] : CONFLICT_PAIRS[i][0];
        if (digitalRead(RELAY_PINS[other]) == LOW) {
          digitalWrite(RELAY_PINS[other], HIGH);
          Serial.print("  X Cleared: "); Serial.println(DIR_NAMES[other]);
        }
        break;
      }
    }
    digitalWrite(RELAY_PINS[d], LOW);
    Serial.print("  > "); Serial.print(DIR_NAMES[d]); Serial.println(" ON");
  } else {
    digitalWrite(RELAY_PINS[d], HIGH);
    Serial.print("  | "); Serial.print(DIR_NAMES[d]); Serial.println(" OFF");
  }
}

// ============ WEBSOCKET EVENT ============
void webSocketEvent(WStype_t type, uint8_t* payload, size_t length) {
  switch (type) {
    case WStype_CONNECTED:
      Serial.println("WS: Connected");
      wsConnected = true;
      {
        JsonDocument doc;
        doc["type"]       = "register";
        doc["clientType"] = "pcb";
        doc["deviceId"]   = DEVICE_ID;
        String msg;
        serializeJson(doc, msg);
        wsClient.sendTXT(msg);
        Serial.print("Registered as ["); Serial.print(DEVICE_ID); Serial.println("]");
      }
      break;

    case WStype_DISCONNECTED:
      Serial.println("WS: Disconnected");
      wsConnected = false;
      stopAll();
      break;

    case WStype_TEXT:
      {
        JsonDocument doc;
        DeserializationError err = deserializeJson(doc, payload, length);
        if (!err && doc["d"].is<int>()) {
          int d = doc["d"];
          int s = doc["s"];
          Serial.print("[CMD] d="); Serial.print(d);
          Serial.print(" s="); Serial.println(s);
          handleCommand(d, s);
        }
      }
      break;

    case WStype_ERROR:
      Serial.println("WS: Error");
      break;

    default: break;
  }
}

// ============ WIFI ============
bool connectWifi() {
  Serial.print("Connecting to WiFi");
  WiFi.begin(WIFI_SSID, WIFI_PASSWORD);
  int timeout = 20;
  while (WiFi.status() != WL_CONNECTED && timeout-- > 0) {
    delay(500); Serial.print(".");
  }
  if (WiFi.status() == WL_CONNECTED) {
    Serial.println();
    Serial.print("WiFi! IP: "); Serial.println(WiFi.localIP());
    return true;
  }
  Serial.println("\nWiFi failed.");
  return false;
}

void checkWifi() {
  if (millis() - lastWifiCheck < WIFI_CHECK_INTERVAL) return;
  lastWifiCheck = millis();
  if (WiFi.status() != WL_CONNECTED) {
    Serial.println("WiFi lost! Reconnecting...");
    wsConnected = false;
    stopAll();
    WiFi.disconnect();
    delay(1000);
    connectWifi();
  }
}

// ============ SETUP ============
void setup() {
  Serial.begin(115200);
  delay(2000);

  Serial.println("========================================");
  Serial.print("  Spray Bot - Receiver ["); Serial.print(DEVICE_ID); Serial.println("]");
  Serial.println("========================================\n");

  for (int i = 0; i < DIR_COUNT; i++) {
    digitalWrite(RELAY_PINS[i], HIGH);  // relays OFF before setting as output
    pinMode(RELAY_PINS[i], OUTPUT);
  }

  if (!connectWifi()) {
    while (true) delay(1000);
  }

  wsClient.beginSSL(WS_HOST, WS_PORT, WS_PATH, (uint8_t*)NULL);
  wsClient.onEvent(webSocketEvent);
  wsClient.setReconnectInterval(5000);
  wsClient.enableHeartbeat(25000, 5000, 3);

  Serial.println("Connecting to WebSocket...");
}

// ============ LOOP ============
void loop() {
  checkWifi();
  wsClient.loop();
}
